### ISSUE TITLE
Bug: Strange interactions between write_params and set_pfiles

### ISSUE THREAD

**Issue by hamogu** [cxcsds/ciao-contrib#448]
Using the `runtool` I can change the parameters of any tool and write that back to a parameter file, using a default name, any random name, or a name that's like the tool.
```
import ciao_contrib.runtool as rt

rt.dmcopy.clobber = True
rt.dmcopy.write_params()
rt.dmcopy.write_params('mytest.par')
rt.dmcopy.write_params('dmcopy.par')
```
However, if I first set the environment to the current dir, the last one will fail.
```
import os
rt.set_pfiles(os.path.abspath(os.curdir))
rt.dmcopy.write_params()
rt.dmcopy.write_params('mytest.par')
rt.dmcopy.write_params('dmcopy.par')

---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-8-fee1add62e19> in <module>
----> 1 rt.dmcopy.write_params('dmcopy.par')

~/anaconda3/envs/ciao/lib/python3.7/site-packages/ciao_contrib/runtool.py in write_params(self, parfile)
   1491             if not pname.endswith(".par"):
   1492                 pname += ".par"
-> 1493             self._create_parfile_copy(pname)
   1494 
   1495         stackfiles = self._update_parfile(pname)

~/anaconda3/envs/ciao/lib/python3.7/site-packages/ciao_contrib/runtool.py in _create_parfile_copy(self, parfile)
   1142 
   1143         try:
-> 1144             ofile = pio.paramgetpath(toolname)
   1145             v5("Copying par file {0} to {1}".format(ofile, parfile))
   1146             with open(ofile, "r") as ifh:

OSError: paramgetpath() Could not open parameter file
```
That's somewhat surprising because not giving a file name for the parameter file uses a default that's called "dmcopy.par". (That also makes for a simple work-around: Just do not pass in a name.)

I've tried to debug this, but it's a little confusing because the error gets thrown inside a try-except block. If I enter the debugger at that point, I can execute it line by line and it works just fine, but it always fails if called as above. (I execute this code in a jupyter notebook, but that should not be relevant here.)

**Comment by DougBurke** [comment]
The `set_pfiles` is a red-herring: it's because you are in the same directory as the parameter files - e.g.

```
>>> import paramio as pio
>>> pio.paramgetpath('dmcopy')
'/home/dburke/cxcds_param4/dmcopy.par'
>>> import os
>>> os.chdir('/home/dburke/cxcds_param4')
>>> pio.paramgetpath('dmcopy')
'/home/dburke/cxcds_param4/dmcopy.par'
>>> import ciao_contrib.runtool as rt
>>> rt.dmcopy.write_params('dmcopy.par')
Error in parameter file: line 1: field Name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dburke/anaconda/envs/ciao413/lib/python3.8/site-packages/ciao_contrib/runtool.py", line 1493, in write_params
    self._create_parfile_copy(pname)
  File "/home/dburke/anaconda/envs/ciao413/lib/python3.8/site-packages/ciao_contrib/runtool.py", line 1144, in _create_parfile_copy
    ofile = pio.paramgetpath(toolname)
OSError: paramgetpath() Could not open parameter file
```

Now, why this is a problem is a different question, since the following works:

```
>>> import paramio as pio
>>> pio.paramgetpath('dmcopy')
'/home/dburke/cxcds_param4/dmcopy.par'
>>> import os
>>> os.chdir('/home/dburke/cxcds_param4')
>>> pio.paramgetpath('dmcopy')
'/home/dburke/cxcds_param4/dmcopy.par'
```

**Comment by kglotfelty** [comment]
FYI - there is some wonkiness with the parameter interface (sometimes you do need to specify eg `./dmcopy.par`  to get things to work -- I forget the exact circumstances).

But I think in this case, `_create_parfile_copy` is hitting a race condition

https://github.com/cxcsds/ciao-contrib/blob/83ae83e8cb55b97770d74ed1e52cc194015ced1e/ciao_contrib/runtool.py#L1141-L1146

I think in this case, `ofile` and `parfile` are both just `dmcopy.par`, so the `open()` is trashing the existing `dmcopy.par` before it can be copied over-itself.

Maybe could check if `abspath(ofile) == abspath(parfile)`

**Comment by DougBurke** [comment]
So, `write_params` begins with

https://github.com/cxcsds/ciao-contrib/blob/a2f428808ea31d5c8b42c9e601621cfc1b98d75f/ciao_contrib/runtool.py#L1518-L1525

which means that it treates `write_params('dmcopy.par')` and `write_params()` [for `dmcopy`] differently, which is why the version without an argument works. Writing tests for this is likely harder than fixing it...

**Comment by DougBurke** [comment]
Oh darn it. There's a subtle difference between using a random name like `mytest.par` and the name of an existing parameter file, like `dmcopy.par`. For the former we essentially copy the tool's .par file to `mytest.par` and then we can modify it in place. For any name which matches an existing .par file then we end up using the existing one, even if we were to copy it to the current directory.  So the semantics of `write_params` is not well specified. In particular, do we want the file to be in the current working directory OR does it not really matter, as long as the parameter tools can access/use it?

One thing I can maybe do is treat `rt.dmcopy.write_params('dmcopy.par')` and `rt.dmcopy.write_params()` to be the same.

**Comment by hamogu** [comment]
It's been a while and I did not write down how I discovered this originally. (It's a perfect example of a developer's bug report: Minimal example code, reproducible, beginnings of code spelunking, but no indication why this bug matters at all.) If I remember correctly, I wanted to write a parameter file into the same directory where I have the data, so that I would end up with one parameter file for each OBSID. Later, I can then re-run certain things and pick up exactly that parameter file
These files would be different each each OBSID, so putting them into the exiting global place does not work. 

I also did not use dmcopy, but something more complicated. I just found that dmcopy can be used to make a minimal example of the problem. I think it was ardlib.par, where I want to set the badpix file to the right one for that specific OBSID and then have later scripts pick that up; it's particularly relevant there because I can't set the ardlib parameters at command line parameters for e.g. specextract.
